File: clients/openai-go/tests/extra_content_test.go
/**
Tests for `tensorzero_extra_content` round-trip support.

These tests verify that extra content blocks (Thought, Unknown) can be:
1. Received from the API in responses
2. Sent back to the API in follow-up requests (round-trip)
*/

package tests

import (
	"encoding/json"
	"testing"

	"github.com/google/uuid"
	"github.com/openai/openai-go/v3"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestExtraContent(t *testing.T) {
	t.Run("should round-trip extra content non-streaming", func(t *testing.T) {
		episodeID, _ := uuid.NewV7()

		// Step 1: Make inference request with a model that returns Thought content
		// The dummy::reasoner model returns [Thought, Text] content
		messages := []openai.ChatCompletionMessageParamUnion{
			openai.UserMessage("Hello"),
		}

		req := &openai.ChatCompletionNewParams{
			Model:    "tensorzero::model_name::dummy::reasoner",
			Messages: messages,
		}
		req.SetExtraFields(map[string]any{
			"tensorzero::episode_id": episodeID.String(),
		})

		resp, err := client.Chat.Completions.New(ctx, *req)
		require.NoError(t, err, "Initial API request failed")

		// Step 2: Verify response structure
		require.NotEmpty(t, resp.Choices, "Response should have choices")
		content := resp.Choices[0].Message.Content
		require.NotEmpty(t, content, "Response should have content")

		// Check for tensorzero_extra_content at the message level
		extraContentField, ok := resp.Choices[0].Message.JSON.ExtraFields["tensorzero_extra_content"]
		require.True(t, ok, "Response message should have tensorzero_extra_content field")

		var extraContent []map[string]interface{}
		err = json.Unmarshal([]byte(extraContentField.Raw()), &extraContent)
		require.NoError(t, err, "Failed to parse tensorzero_extra_content")
		require.NotEmpty(t, extraContent, "Extra content should have at least one block")

		// Verify the structure of the thought block
		thoughtBlock := extraContent[0]
		assert.Equal(t, "thought", thoughtBlock["type"], "First block should be a thought")
		assert.NotNil(t, thoughtBlock["insert_index"], "Thought block should have insert_index")
		assert.NotNil(t, thoughtBlock["text"], "Thought block should have text field")

		// Step 3: Round-trip - send the extra content back as an assistant message
		roundtripReq := &openai.ChatCompletionNewParams{
			Model: "tensorzero::model_name::dummy::echo",
			Messages: []openai.ChatCompletionMessageParamUnion{
				openai.UserMessage("Hello"),
				openai.AssistantMessage(content),
				openai.UserMessage("Continue"),
			},
		}

		// Set extra content on the assistant message via extra fields
		assistantMsg := roundtripReq.Messages[1].OfAssistant()
		if assistantMsg != nil {
			assistantMsg.SetExtraFields(map[string]any{
				"tensorzero_extra_content": extraContent,
			})
		}

		roundtripReq.SetExtraFields(map[string]any{
			"tensorzero::episode_id": episodeID.String(),
		})

		roundtripResp, err := client.Chat.Completions.New(ctx, *roundtripReq)
		require.NoError(t, err, "Round-trip API request failed")

		// Verify round-trip succeeded
		require.NotEmpty(t, roundtripResp.Choices, "Round-trip response should have choices")
	})

	t.Run("should round-trip extra content streaming", func(t *testing.T) {
		episodeID, _ := uuid.NewV7()

		// Step 1: Make streaming inference request
		messages := []openai.ChatCompletionMessageParamUnion{
			openai.UserMessage("Hello"),
		}

		req := &openai.ChatCompletionNewParams{
			Model:    "tensorzero::model_name::dummy::reasoner",
			Messages: messages,
		}
		req.SetExtraFields(map[string]any{
			"tensorzero::episode_id": episodeID.String(),
		})

		stream := client.Chat.Completions.NewStreaming(ctx, *req)
		require.NotNil(t, stream, "Streaming response should not be nil")

		// Step 2: Collect chunks and extract extra content
		var extraContentChunks []map[string]interface{}
		var contentText string

		for stream.Next() {
			chunk := stream.Current()
			if len(chunk.Choices) > 0 {
				delta := chunk.Choices[0].Delta

				// Collect text content
				if delta.Content != "" {
					contentText += delta.Content
				}

				// Collect extra content chunks from delta
				extraContentField, ok := delta.JSON.ExtraFields["tensorzero_extra_content"]
				if ok {
					var chunkExtraContent []map[string]interface{}
					err := json.Unmarshal([]byte(extraContentField.Raw()), &chunkExtraContent)
					if err == nil {
						extraContentChunks = append(extraContentChunks, chunkExtraContent...)
					}
				}
			}
		}
		require.NoError(t, stream.Err(), "Stream encountered an error")

		// Step 3: Verify we received extra content in streaming
		assert.NotEmpty(t, extraContentChunks, "Streaming should include extra content chunks")

		// Reconstruct extra content for round-trip (filter for chunks with insert_index)
		var reconstructedExtraContent []map[string]interface{}
		for _, chunk := range extraContentChunks {
			if chunk["insert_index"] != nil {
				reconstructedExtraContent = append(reconstructedExtraContent, chunk)
			}
		}

		// Step 4: Round-trip if we have valid content
		if len(reconstructedExtraContent) > 0 && len(contentText) > 0 {
			roundtripReq := &openai.ChatCompletionNewParams{
				Model: "tensorzero::model_name::dummy::echo",
				Messages: []openai.ChatCompletionMessageParamUnion{
					openai.UserMessage("Hello"),
					openai.AssistantMessage(contentText),
					openai.UserMessage("Continue"),
				},
			}

			// Set extra content on the assistant message via extra fields
			assistantMsg := roundtripReq.Messages[1].OfAssistant()
			if assistantMsg != nil {
				assistantMsg.SetExtraFields(map[string]any{
					"tensorzero_extra_content": reconstructedExtraContent,
				})
			}

			roundtripReq.SetExtraFields(map[string]any{
				"tensorzero::episode_id": episodeID.String(),
			})

			roundtripResp, err := client.Chat.Completions.New(ctx, *roundtripReq)
			require.NoError(t, err, "Streaming round-trip API request failed")

			// Verify round-trip succeeded
			assert.NotEmpty(t, roundtripResp.Choices, "Streaming round-trip response should have choices")
		}
	})
}

File: clients/openai-node/tests/extra-content.test.ts
/**
 * Tests for `tensorzero_extra_content` round-trip support.
 *
 * These tests verify that extra content blocks (Thought, Unknown) can be:
 * 1. Received from the API in responses
 * 2. Sent back to the API in follow-up requests (round-trip)
 */
import { describe, it, expect, beforeAll } from "vitest";
import OpenAI from "openai";
import { ChatCompletionMessageParam } from "openai/resources";
import { v7 as uuidv7 } from "uuid";

let client: OpenAI;

beforeAll(() => {
  client = new OpenAI({
    apiKey: "donotuse",
    baseURL: "http://127.0.0.1:3000/openai/v1",
  });
});

interface ExtraContentBlock {
  type: "thought" | "unknown";
  insert_index?: number;
  text?: string;
  signature?: string | null;
  summary?: string | null;
  provider_type?: string | null;
  data?: unknown;
  model_name?: string | null;
  provider_name?: string | null;
}

describe("Extra Content", () => {
  it("should round-trip extra content non-streaming", async () => {
    const episodeId = uuidv7();

    // Step 1: Make inference request with a model that returns Thought content
    // The dummy::reasoner model returns [Thought, Text] content
    const messages: ChatCompletionMessageParam[] = [
      { role: "user", content: "Hello" },
    ];

    const result = await client.chat.completions.create({
      messages,
      model: "tensorzero::model_name::dummy::reasoner",
      // @ts-expect-error - custom TensorZero property
      "tensorzero::episode_id": episodeId,
    });

    // Step 2: Verify response structure
    expect(result.choices[0].message.content).toBeDefined();

    const extraContent: ExtraContentBlock[] | undefined = (
      result.choices[0].message as unknown as {
        tensorzero_extra_content?: ExtraContentBlock[];
      }
    ).tensorzero_extra_content;
    expect(extraContent).toBeDefined();
    expect(Array.isArray(extraContent)).toBe(true);
    expect(extraContent!.length).toBeGreaterThan(0);

    // Verify the structure of the thought block
    const thoughtBlock = extraContent![0];
    expect(thoughtBlock.type).toBe("thought");
    expect(thoughtBlock.insert_index).toBeDefined();
    expect(thoughtBlock.text).toBeDefined();

    // Step 3: Round-trip - send the extra content back as an assistant message
    const roundtripMessages: ChatCompletionMessageParam[] = [
      { role: "user", content: "Hello" },
      {
        role: "assistant",
        content: result.choices[0].message.content!,
        // @ts-expect-error - custom TensorZero property
        tensorzero_extra_content: extraContent,
      },
      { role: "user", content: "Continue" },
    ];

    const roundtripResult = await client.chat.completions.create({
      messages: roundtripMessages,
      model: "tensorzero::model_name::dummy::echo",
      // @ts-expect-error - custom TensorZero property
      "tensorzero::episode_id": episodeId,
    });

    // Verify round-trip succeeded
    expect(roundtripResult.choices[0].message).toBeDefined();
  });

  it("should round-trip extra content streaming", async () => {
    const episodeId = uuidv7();

    // Step 1: Make streaming inference request
    const messages: ChatCompletionMessageParam[] = [
      { role: "user", content: "Hello" },
    ];

    // @ts-expect-error - custom TensorZero property
    const stream = await client.chat.completions.create({
      messages,
      model: "tensorzero::model_name::dummy::reasoner",
      stream: true,
      "tensorzero::episode_id": episodeId,
    });

    // Step 2: Collect chunks and extract extra content
    const extraContentChunks: ExtraContentBlock[] = [];
    let contentText = "";

    for await (const chunk of stream) {
      const delta = chunk.choices[0]?.delta;
      if (delta) {
        // Collect text content
        if (delta.content) {
          contentText += delta.content;
        }

        // Collect extra content chunks
        const extraContent: ExtraContentBlock[] | undefined = (
          delta as unknown as {
            tensorzero_extra_content?: ExtraContentBlock[];
          }
        ).tensorzero_extra_content;
        if (extraContent) {
          extraContentChunks.push(...extraContent);
        }
      }
    }

    // Step 3: Verify we received extra content in streaming
    expect(extraContentChunks.length).toBeGreaterThan(0);

    // Reconstruct extra content for round-trip (filter for chunks with insert_index)
    const reconstructedExtraContent = extraContentChunks.filter(
      (chunk) => chunk.insert_index !== undefined
    );

    // Step 4: Round-trip if we have valid content
    if (reconstructedExtraContent.length > 0 && contentText.length > 0) {
      const roundtripMessages: ChatCompletionMessageParam[] = [
        { role: "user", content: "Hello" },
        {
          role: "assistant",
          content: contentText,
          // @ts-expect-error - custom TensorZero property
          tensorzero_extra_content: reconstructedExtraContent,
        },
        { role: "user", content: "Continue" },
      ];

      const roundtripResult = await client.chat.completions.create({
        messages: roundtripMessages,
        model: "tensorzero::model_name::dummy::echo",
        // @ts-expect-error - custom TensorZero property
        "tensorzero::episode_id": episodeId,
      });

      // Verify round-trip succeeded
      expect(roundtripResult.choices[0].message).toBeDefined();
    }
  });
});
